lightning/src/ln/channel.rs

```diff
@@ -8205,6 +8667,71 @@ impl<SP: Deref> InboundV2Channel<SP> where SP::Target: SignerProvider {
 	pub fn get_accept_channel_v2_message(&self) -> msgs::AcceptChannelV2 {
 		self.generate_accept_channel_v2_message()
 	}
+
+	pub fn begin_interactive_funding_tx_construction<ES: Deref>(&mut self, signer_provider: &SP,
+		entropy_source: &ES, holder_node_id: PublicKey, funding_inputs: Vec<(TxIn, TransactionU16LenLimited)>
+	) -> Result<Option<InteractiveTxMessageSend>, APIError>
+	where ES::Target: EntropySource
+	{
+		self.context.begin_interactive_funding_tx_construction(&self.dual_funding_context,
+			signer_provider, entropy_source, holder_node_id, false /* is_initiator */, funding_inputs)
+	}
+
+	pub fn funding_tx_constructed<L: Deref>(
+		mut self, counterparty_node_id: &PublicKey, mut signing_session: InteractiveTxSigningSession, signer_provider: &SP, logger: &L
+	) -> Result<(Channel<SP>, msgs::CommitmentSigned, Option<Event>), (Self, ChannelError)>
+	where
+		L::Target: Logger
+	{
+		let mut output_index = None;
+		let expected_spk = self.context.get_funding_redeemscript().to_v0_p2wsh();
+		for (idx, outp) in signing_session.constructed_transaction.output.iter().enumerate() {
+			if outp.script_pubkey == expected_spk && outp.value == self.context.get_value_satoshis() {
+				if output_index.is_some() {
+					return Err((self, ChannelError::Close("Multiple outputs matched the expected script and value".to_owned())));
+				}
+				output_index = Some(idx as u16);
+			}
+		}
+		if output_index.is_none() {
+			if output_index.is_some() {
```

Comment:
These two conditions are contradictory, inner will never evaulate, please check

Reply:
This is on my radar. I'm waiting on #2419 to land before a bunch of fixups here.